packages/core/src/Site/index.js

```diff
@@ -696,90 +713,107 @@ class Site {
      hence we need to take this into account.
      */
-    const regeneratePagesBeingViewed = uniqueUrls.map((normalizedUrl) => {
+    const regeneratePagesBeingViewed = uniqueUrls.map(async (normalizedUrl) => {
       this._setTimestampVariable();
       const pageToRebuild = this.pages.find(page =>
         FsUtil.removeExtension(page.pageConfig.sourcePath) === normalizedUrl);
 
       if (!pageToRebuild) {
-        return Promise.resolve();
+        return;
       }
 
       this.toRebuild.delete(normalizedUrl);
-      return pageToRebuild.generate(this.externalManager)
-        .then(() => this.writeSiteData())
-        .then(() => {
-          const endTime = new Date();
-          const totalBuildTime = (endTime - startTime) / 1000;
-          logger.info(`Lazy website regeneration complete! Total build time: ${totalBuildTime}s`);
-        })
-        .catch(error => Site.rejectHandler(error, [this.tempPath, this.outputPath]));
+      try {
+        await pageToRebuild.generate(this.externalManager);
+        await this.writeSiteData();
+        Site.calculateBuildTimeForRebuildPageBeingViewed(startTime);
+      } catch (error) {
+        await Site.rejectHandler(error, [this.tempPath, this.outputPath]);
+      }
     });
 
-    return Promise.all(regeneratePagesBeingViewed)
-      .then(() => fs.remove(this.tempPath));
+    await Promise.all(regeneratePagesBeingViewed);
+    await fs.remove(this.tempPath);
   }
 
-  _rebuildSourceFiles() {
+  /**
+   * Helper function for _rebuildPageBeingViewed().
+   */
+  static calculateBuildTimeForRebuildPageBeingViewed(startTime) {
+    const endTime = new Date();
+    const totalBuildTime = (endTime - startTime) / 1000;
+    return logger.info(`Lazy website regeneration complete! Total build time: ${totalBuildTime}s`);
+  }
+
+  async _rebuildSourceFiles() {
     logger.info('Page added or removed, updating list of site\'s pages...');
     this.beforeSiteGenerate();
 
     this.layoutManager.removeLayouts();
 
     const removedPageFilePaths = this.updateAddressablePages();
-    return this.removeAsset(removedPageFilePaths)
-      .then(() => {
-        if (this.onePagePath) {
-          this.mapAddressablePagesToPages(this.addressablePages || [], this.getFavIconUrl());
+    try {
+      await this.removeAsset(removedPageFilePaths);
+      await this.rebuildRequiredPages();
+    } catch (error) {
+      await Site.rejectHandler(error, [this.tempPath, this.outputPath]);
+    }
+  }
 
-          return this.rebuildPageBeingViewed(this.currentPageViewed)
-            .then(() => this.lazyBuildAllPagesNotViewed());
-        }
+  /**
+   * Helper function for _rebuildSourceFiles().
+   */
+  async rebuildRequiredPages() {
+    if (this.onePagePath) {
+      this.mapAddressablePagesToPages(this.addressablePages || [], this.getFavIconUrl());
+
+      await this.rebuildPageBeingViewed(this.currentPageViewed);
+      await this.lazyBuildAllPagesNotViewed();
```

Comment:
@jonahtanjz just noticed this, let's re-add the missing `return` termination sometime